feat: support new Bootstrapped and Errored watch events

smira opened the pull request:
In the regular COSI, new event types are enabled immediately.

With protobuf client, new types are enabled when using new COSI client, old COSI client against new server won't receive new events, as old implementation panics on nil Resource in the event structure.

Fixes #175

Signed-off-by: Andrey Smirnov <andrey.smirnov@talos-systems.com>

smira commented:
See https://github.com/cosi-project/specification/pull/5

utkuozdemir commented on pkg/state/impl/inmem/collection.go:
```diff
@@ -455,10 +465,15 @@ func (collection *ResourceCollection) WatchAll(ctx context.Context, ch chan<- st
 
+				select {
+				case ch <- state.Event{
+					Type: state.Errored,
```
why don't we fill the error field when the state is errored?

smira replied:
:facepalm:  thanks for noticing that, I will write a test for it!

utkuozdemir replied:
I think the assertions need to be inverted as well :)

utkuozdemir reviewed (approved):
🆒

smira commented:
I addressed the feedback, and added tests for Watch errors (simulated buffer overrun), and also controller-runtime now fails if any internal watch call fails (also with a unit-test)

smira commented:
/m
